packages/gatsby/src/commands/build.ts

```diff
@@ -232,14 +238,34 @@ module.exports = async function build(program: IBuildArgs): Promise<void> {
   try {
     const result = await buildRenderer(program, Stage.BuildHTML, buildSpan)
     pageRenderer = result.rendererPath
+    if (_CFLAGS_.GATSBY_MAJOR === `4` && shouldGenerateEngines()) {
+      // for now copy page-render to `.cache` so page-ssr module can require it as a sibling module
+      const outputDir = path.join(program.directory, `.cache`, `page-ssr`)
+      engineBundlingPromises.push(
+        fs
+          .ensureDir(outputDir)
+          .then(() =>
+            fs.copyFile(
+              result.rendererPath,
+              path.join(outputDir, `render-page.js`)
+            )
+          )
+      )
+    }
     waitForCompilerCloseBuildHtml = result.waitForCompilerClose
   } catch (err) {
     buildActivityTimer.panic(structureWebpackErrors(Stage.BuildHTML, err))
   } finally {
     buildSSRBundleActivityProgress.end()
   }
 
+  if (_CFLAGS_.GATSBY_MAJOR === `4` && shouldGenerateEngines()) {
+    // well, tbf we should just generate this in `.cache` and avoid deleting it :shrug:
```

Comment:
👍 I assume doing this in v3 will be a breaking change?

Reply:
Possibly yeah, I just didn't want to introduce more changes in this PR than absolutely needed.